KOGITO-7965: Create a NPM package consolidating test helpers from vscode-extension-* integration tests

LuboTerifaj opened the pull request:
https://issues.redhat.com/browse/KOGITO-7965

domhanak reviewed (commented):
2 questions right away, will check the code later.

1. Is the Typescript upgrade necessary to be done here?

2. Did you check the use of locators pattern from https://github.com/redhat-developer/vscode-extension-tester/tree/master/locators ?

LuboTerifaj commented:
Hello @domhanak, 

unfortunately it is necessary to upgrade typescript version at least in affected packages. There was an issue with `Workbench` initialization due to different versions of typescript within the monorepo. There is a minimal version required by vscode-extension-tester. It is also better to align the typescript versions in the whole monorepo to avoid possible issues.

Yes, I have seen it, but I don't see any added value here. If we want to have it, I would suggest to open a separate jira for it.

It potentially might make sense to use this approach in case we would create a separate "helper" library including all helper classes and methods from specific editors like swf, dashbuilder, bpmn, dmn, pmml, scesim and yard. However, for now I don't see the reason to separate code from tests since it doesn't solve any issues as the code is different for each editor.
Of course, once we have this PR merged, we can decide if want to move there also other non-common parts and make the vscode-extension-common-test-helpers package a library.

Thank you!

LuboTerifaj commented:
I am fixing the errors caused by typescript upgrade. Making the PR to a draft.

LuboTerifaj commented:
I will do the typescript update in a separate jira: https://issues.redhat.com/browse/KOGITO-9133. Once it is merged I will update this PR.

LuboTerifaj commented:
Here is the pull request with typescript version update:

- https://github.com/kiegroup/kie-tools/pull/1622

Those changes are currently part of this PR as well. Once it is merged, I will rebase this pull request if necessary.

LuboTerifaj commented:
Hello @paulovmr , @domhanak ,

This pull-request should be up to date and ready for review.

I rebased the PR with main. It should pass since it already passed once with the same code before I rebased it.

Thanks!

domhanak commented on packages/vscode-extension-common-test-helpers/src/VSCodeTestHelper.ts:
```diff
@@ -127,21 +186,26 @@ export default class VSCodeTestHelper {
-    const webviewLeft = new WebView(editorGroups[0], By.linkText(fileName));
-    const webviewRight = new WebView(editorGroups[1], By.linkText(fileName));
+  private isKieEditorWithTwoWebViews(fileName: string): boolean {
+    return /\.sw\.(json|ya?ml)|\.dash\.ya?ml|\.yard\.ya?ml$/.test(fileName);
```
I feel like this could be hard to maintain if we add more, maybe we add some constant for this? Or a provider that will give us these? `KieSupportedFileExtensions.getSupportedDualViewExtensions()` simirally for single view one

LuboTerifaj replied:
Are you aware of any provider that I can use? If not I can move it to a constant within the package.

LuboTerifaj replied:
Or did you mean to create a new KieSupportedFileExtensions file in current package? I can do that as well.

LuboTerifaj replied:
I have added the commit with moving the extensions to a separate file.

domhanak reviewed (changes requested):
Looks good thank you. Great job

 I think you need to revert the copyrights dates, and there are couple of suggestion that can be addressed but nothing major. Let me know

LuboTerifaj commented on packages/envelope-bus/src/common/EnvelopeBusMessageManager.ts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates.
```
Actually I updated the file in the PR with typescript update and this file was affected, but I forgot to split the license header commit.

LuboTerifaj commented on packages/i18n/src/core/immutableDeepMerge.ts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates.
```
Actually I updated the file in the PR with typescript update and this file was affected, but I forgot to split the license header commit.

LuboTerifaj commented on packages/kie-editors-dev-vscode-extension/it-tests/extension-editors-bpmn.test.ts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates.
```
Actually I updated the file in the PR with typescript update and this file was affected, but I forgot to split the license header commit.

LuboTerifaj replied:
This file was changed, so I think the license header can be updated as well.

LuboTerifaj commented on packages/kie-editors-dev-vscode-extension/it-tests/extension-editors-bpmn.test.ts:
```diff
@@ -122,27 +123,29 @@ describe("KIE Editors Integration Test Suite - BPMN Editor", () => {
-    await testHelper.switchWebviewToFrame(webview);
+    const editorWebviews = await testHelper.openFileFromSidebar("ReuseDataTypeTest.bpmn");
+    webview = editorWebviews[0];
     const bpmnEditorTester = new BpmnEditorTestHelper(webview);
```
`webiew` is actually the variable used within the whole test suite. I would like to avoid changing the whole editor's  test suite implementation style.

domhanak replied:
Isn't it only used to do `webview.switchBack()` after the test? it seems redundant, but I leave it up to you.

LuboTerifaj replied:
It is taken as a parameter in many lines of this file, so I would say it is better to have it in a separate variable. I keep it as it is. Thanks!

LuboTerifaj commented on packages/kie-editors-dev-vscode-extension/it-tests/extension-editors-bpmn.test.ts:
```diff
@@ -173,15 +176,16 @@ describe("KIE Editors Integration Test Suite - BPMN Editor", () => {
-    webview = await testHelper.openFileFromSidebar(MULTIPLE_INSTANCE_BPMN);
-    await testHelper.switchWebviewToFrame(webview);
+    const editorWebviews = await testHelper.openFileFromSidebar(MULTIPLE_INSTANCE_BPMN);
+    webview = editorWebviews[0];
```
`webiew` is actually the variable used within the whole test suite. I would like to avoid changing the whole editor's  test suite implementation style.

LuboTerifaj commented on packages/kie-editors-dev-vscode-extension/it-tests/extension-editors-bpmn.test.ts:
```diff
@@ -242,14 +246,15 @@ describe("KIE Editors Integration Test Suite - BPMN Editor", () => {
-    await testHelper.switchWebviewToFrame(webview);
+    const editorWebviews = await testHelper.openFileFromSidebar(USER_TASK_BPMN);
+    webview = editorWebviews[0];
     const bpmnEditorTester = new BpmnEditorTestHelper(webview);
```
`webiew` is actually the variable used within the whole test suite. I would like to avoid changing the whole editor's  test suite implementation style.

LuboTerifaj commented:
Hello @domhanak , I think there is only one your suggestion that may be needed to change. Could you please have a look at my comments above? Thanks!

domhanak commented on packages/kie-editors-dev-vscode-extension/it-tests/extension-editors-dmn.test.ts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates.
```
Nah, there are conflicting views on this I guess, I was under the impression we  should not update them, but I think we can ignore this.

LuboTerifaj replied:
Yeah, there is no ideal approach for that, but I don't see the issue in updating the year here. I keep the change as it is in my commit. Thanks!

LuboTerifaj commented:
Hello @paulovmr , could you please have a look at this PR? It's already approved from QE. Thank you!

LuboTerifaj commented:
Hello @paulovmr , once approved, please squash the commits before merging. You don't need to preserve the commit history. Only the main commit message should be kept. Thanks!
